File: source/yeon/8주차/code/10844쉬운계단수.cpp
// 백준 10844 쉬운계단수
// cpp solved by dp 
#include<iostream>
#define MOD 1000000000
int main(void){
    // just assign 2x12 arr 
    int N, arr[2][12]={0,};
    scanf("%d", &N);
    int sum = 9;
    
    for(int i = 2; i <11; i++){
        arr[1][i] = 1;
    }

    for(int i = 2; i <= N; i++){
        sum = 0;
        // by using modular calculation, don't have to assign 11x102 size of arr
        // moving up & down so that decrease use of space complexity
        for(int j = 1; j < 11; j++){
            arr[i%2][j] = (arr[(i-1)%2][j-1]+arr[(i-1)%2][j+1]) % MOD;
            sum = (sum + arr[i%2][j]) % MOD;
        }
    }
    printf("%d",sum);
    return 0;
}

File: source/yeon/8주차/code/10844쉬운계단수.py
# 백준 10844 쉬운계단수
# python solved by dp
N = int(input())
MOD, _sum = 1000000000, 0

# assign 11x101 dp arr 
dp = [[0 for _ in range(11)]for _ in range(101)]

# in case of one digit
for i in range(1,10):
    dp[0][i] = 1

# we can find out that 
# dp[i][j] = dp[i-1][j-1] + dp[i-1][j+1]
for i in range(1,N+1):
    dp[i][0] = dp[i-1][1]
    dp[i][10] = 0
    for j in range(10):
        dp[i][j] = (dp[i-1][j-1] + dp[i-1][j+1]) % MOD

for i in range(10):
    _sum = (_sum + dp[N-1][i]) % MOD 
print(_sum)